src/solvers/adaptive_time_solver.C

```diff
@@ -121,7 +131,54 @@ void AdaptiveTimeSolver::advance_timestep ()
      _system.get_dof_map().get_send_list());
 }
 
+void AdaptiveTimeSolver::adjoint_advance_timestep ()
+{
+  // All calls to adjoint_advance_timestep are made in the user's
+  // code. This first call is made immediately after the adjoint initial conditions
+  // are set. This is in the user code outside the adjoint time loop.
+  if (!first_adjoint_step)
+    {
+      // The adjoint system has been solved. We need to store the adjoint solution and
+      // load the primal solutions for the next time instance (t - delta_ti).
+      _system.time -= this->completedtimestep_deltat;
+    }
+  else
+    {
+      // After setting the initial conditions, we have to retrieve the saved primal solutions
+      core_time_solver->get_solution_history().retrieve(true, _system.time);
+
+      // And store the adjoint initial condition, apart from the initial condition these retrieval
+      // and store operations are handled by the core time solver directly
+      core_time_solver->get_solution_history().store(true, _system.time);
 
+      // We also need to tell the core time solver that the adjoint initial conditions have been set
+      core_time_solver->set_first_adjoint_step(false);
+
+      // The first adjoint step simply saves the given adjoint initial condition
+      // So there is a store, but no solve, no actual timestep, so no need to change system time
+      first_adjoint_step = false;
+    }
+
+  // For an adaptive time solver, all solution history operations are handled
+  // by the core time solver.
+
+  // Dont forget to localize the old_nonlinear_solution !
+  _system.get_vector("_old_nonlinear_solution").localize
+    (*old_local_nonlinear_solution,
+    _system.get_dof_map().get_send_list());
+}
+
+void AdaptiveTimeSolver::retrieve_timestep ()
+{
+  // Ask the core time solver to retrieve all the stored vectors
+  // at the current time
+  core_time_solver->retrieve_timestep();
+
+  // Dont forget to localize the old_nonlinear_solution !
+  _system.get_vector("_old_nonlinear_solution").localize
+    (*old_local_nonlinear_solution,
+    _system.get_dof_map().get_send_list());
```

Comment:
If this is necessary shouldn't the core time solver have already done it?  And if not then it's too expensive to do unnecessarily.

Reply:
Yeah, will remove this if not necessary.